server/src/main/java/org/opensearch/indices/IndicesRequestCache.java

```diff
@@ -371,40 +415,301 @@ public int hashCode() {
+         *
+         * @param cleanupKey the CleanupKey that has been marked for cleanup
+         */
+        private void incrementStaleKeysCount(CleanupKey cleanupKey) {
+            if (stalenessThreshold == 0.0 || cleanupKey.entity == null) {
+                return;
+            }
+            IndexShard indexShard = (IndexShard) cleanupKey.entity.getCacheIdentity();
+            if (indexShard == null) {
+                logger.warn("IndexShard is null for CleanupKey: {}", cleanupKey.readerCacheKeyId);
+                return;
+            }
+            ShardId shardId = indexShard.shardId();
+
+            // Using computeIfPresent to atomically operate on the countMap for a given shardId
+            cleanupKeyToCountMap.computeIfPresent(shardId, (key, countMap) -> {
+                if (cleanupKey.readerCacheKeyId == null) {
+                    // Aggregate and add to staleKeysCount atomically if readerCacheKeyId is null
+                    int totalSum = countMap.values().stream().mapToInt(Integer::intValue).sum();
+                    staleKeysCount.addAndGet(totalSum);
+                    // Return null to automatically remove the mapping for shardId
+                    return null;
+                } else {
+                    // Update staleKeysCount based on specific readerCacheKeyId, then remove it from the countMap
+                    countMap.computeIfPresent(cleanupKey.readerCacheKeyId, (k, v) -> {
+                        staleKeysCount.addAndGet(v);
+                        // Return null to remove the key after updating staleKeysCount
+                        return null;
+                    });
+
+                    // Check if countMap is empty after removal to decide if we need to remove the shardId entry
+                    if (countMap.isEmpty()) {
+                        return null; // Returning null removes the entry for shardId
+                    }
+                }
+                return countMap; // Return the modified countMap to keep the mapping
+            });
+        }
+
+        // package private for testing
+        AtomicInteger getStaleKeysCount() {
+            return staleKeysCount;
         }
-        if (!currentKeysToClean.isEmpty() || !currentFullClean.isEmpty()) {
+
+        /**
+         * Clean cache based on stalenessThreshold
+         */
+        void cleanCache() {
+            cleanCache(stalenessThreshold);
+        }
+
+        /**
+         * Force Clean cache without checking stalenessThreshold
+         */
+        private void forceCleanCache() {
+            cleanCache(0);
+        }
+
+        /**
+         * Cleans the cache based on the provided staleness threshold.
+         * <p>If the percentage of stale keys in the cache is less than this threshold,the cache cleanup process is skipped.
+         * @param stalenessThreshold The staleness threshold as a double.
+         */
+        private void cleanCache(double stalenessThreshold) {
```

Comment:
This method was synchronized before why are we changing it ?

Reply:
idk how this creeped through, this should be synchronized. Addressed.